allennlp_rc/models/transformer_qa.py

```diff
@@ -0,0 +1,233 @@
+            self._span_end_accuracy(span_end_logits, span_end, span_mask)
+
+            output_dict["loss"] = loss
+
+        # Compute the EM and F1 on SQuAD and add the tokenized input to the output.
+        if metadata is not None:
+            best_spans = best_spans.detach().cpu().numpy()
+
+            output_dict["best_span_str"] = []
+            context_tokens = []
+            for metadata_entry, best_span in zip(metadata, best_spans):
+                context_tokens_for_question = metadata_entry["context_tokens"]
+                context_tokens.append(context_tokens_for_question)
+
+                best_span -= 1 + len(metadata_entry["question_tokens"]) + 2
+                assert np.all(best_span >= 0)
+
+                predicted_start, predicted_end = tuple(best_span)
+                assert predicted_start >= 0
+                assert predicted_end < len(context_tokens_for_question)
+
+                while (
+                    predicted_start >= 0
+                    and context_tokens_for_question[predicted_start].idx is None
+                ):
+                    predicted_start -= 1
+                if predicted_start < 0:
+                    logger.warning(
+                        f"Could not map the token '{context_tokens_for_question[best_span[0]].text}' at index "
+                        f"'{best_span[0]}' to an offset in the original text. Did you configure your tokenizer "
+                        'with "calculate_character_offsets": true?'
+                    )
+                    character_start = 0
+                else:
+                    character_start = context_tokens_for_question[predicted_start].idx
+
+                while (
+                    predicted_end < len(context_tokens_for_question)
+                    and context_tokens_for_question[predicted_end].idx is None
+                ):
+                    predicted_end += 1
+                if predicted_end >= len(context_tokens_for_question):
+                    logger.warning(
+                        f"Could not map the token '{context_tokens_for_question[best_span[1]].text}' at index "
+                        f"'{best_span[1]}' to an offset in the original text. Did you configure your tokenizer "
+                        'with "calculate_character_offsets": true?'
+                    )
+                    character_end = len(metadata_entry["context"])
+                else:
+                    end_token = context_tokens_for_question[predicted_end]
+                    character_end = end_token.idx + len(sanitize_wordpiece(end_token.text))
+
+                best_span_string = metadata_entry["context"][character_start:character_end]
+                output_dict["best_span_str"].append(best_span_string)
+
+                answers = metadata_entry.get("answers")
+                self._metrics(best_span_string, answers)
+            output_dict["context_tokens"] = context_tokens
+        return output_dict
```

Comment:
There is a fair amount of logic here without tests. This can be tricky with models of course, but I can offer a few possibilities:

1) Use `ensure_model_can_train_save_and_load`. See, e.g., https://github.com/allenai/allennlp/blob/master/allennlp/tests/models/coreference_resolution/coref_test.py#L14. This is quite lightweight and helps catch basic regressions (like zero gradients). Highly recommended.

2) Break out some of the meatier tensor transformations into a function or functions and add unit tests. If you make this function accept embeddings, it's pretty straightforward to give it some fake data and for a reader to be able to verify that it makes sense. The code here is dense enough that I'm having trouble following along. I think this would help a bunch.

Reply:
I put a test for the model into 2e9bb9cd2c4c55c56fb8942aeb15068fcde07558 and some subsequent commits.